### ISSUE TITLE
Smart.Format with multi threading

### ISSUE THREAD

**Issue by karelkral** [axuno/SmartFormat#273]
I have an issue with Smart.Format when using from async code running in parallel. This code is called from multiple threads:
```
string hookUri = Smart.Format(_serviceHookOptions.HookUriTemplate, new { WebRoot = _serviceHookOptions.WebRoot, SecretKey = _apiKeyOptions.ApiKey });
```
and code is randomly ending with this exception:
```
 System.InvalidOperationException: ValueFactory attempted to access the Value property of this instance.

   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.get_Value()
   at SmartFormat.Pooling.SmartPools.FormatPool.get_Instance()
   at SmartFormat.Core.Parsing.Parser.ParseFormat(String inputFormat)
   at SmartFormat.SmartFormatter.Format(IFormatProvider provider, String format, IList`1 args)
   at SmartFormat.SmartFormatter.Format(String format, Object[] args)
   at SmartFormat.Smart.Format(String format, Object arg0)
   at Scia.Cid.AppServer.Tfs.Api.Feat.PullRequestStatus.Services.ServiceHookSubscriptionsManage
```
I suppose that this issue could be caused by some concurrent issues?

**Comment by axunonb** [comment]
Hi Karel,
Yes, you are right.
Please have a look at the Wiki pages
* [Thread Safety](https://github.com/axuno/SmartFormat/wiki/Thread-Safety)
* [SmartSettings](https://github.com/axuno/SmartFormat/wiki/Configuration#smartsettings)

Guess, following these hints will resolve your issue. Please let me know.

**Comment by karelkral** [comment]
Thank you for your help. Issue with System.Lazy was solved, but even with the right settings for thread safety there is another problem with SmartFormat and multiple threads. I have created unit tests for this:

```
using SmartFormat;
using SmartFormat.Core.Settings;
using System.Threading.Tasks;

namespace Scia.Cid.AppServer.Tfs.Api.Tests;

public class SmartFormatTests
{

	/// <summary>
	/// This test is failing with strange parsing errors
	/// </summary>
	[Fact]
	public void MultithreadTest()
	{
		SmartSettings.IsThreadSafeMode = true;

		var range = Enumerable.Range(0, 100);

		Parallel.ForEach(range, (x) =>
		{
			string hookUri = Smart.Format("http://{WebRoot}/tfs/hooks?code={SecretKey}",
				new { WebRoot = "MyWebRoot.cz", SecretKey = x });
			Console.WriteLine(hookUri);
		});
	}

	[Fact]
	public void SingleThreadTest()
	{
		SmartSettings.IsThreadSafeMode = true;

		var range = Enumerable.Range(0, 100);

		foreach (int x in range)
		{
			string hookUri = Smart.Format("http://{WebRoot}/tfs/hooks?code={SecretKey}",
				new { WebRoot = "MyWebRoot.cz", SecretKey = x });
			Console.WriteLine(hookUri);
		}
	}

}
```

Errors are very variable. For example here is one of them:

SmartFormat.Core.Parsing.ParsingErrors: 'The format string has 1 issue:
Format string is missing a closing brace
In: "http://{WebRoot}/tfs/hooks?code={SecretKey}"
At:  -------------------------------------------^ '

**Comment by karelkral** [comment]
When running as part of mutiple test, even SingleThread test was failing with similar exceptions. It is a mystery for me.

**Comment by karelkral** [comment]
Thank you for your help, now it is working. As I can see, this is clearly stated in the documentation :-). 
But as a user of the previous version, I did not read the documentation and got this threading issue with the update to a new SmartFormat.

Thanks for the excellent library.

**Comment by skalahonza** [comment]
thanks for clarification and for this great library
also side note
if you previously used `Smart.Default.Settings.CaseSensitivity = CaseSensitivityType.CaseInsensitive;` don't forget (like I did) to pass the `SmartSettings` instance as parameter

`Smart.CreateDefaultSmartFormat(Smart.Default.Settings).Format(...)`

**Comment by axunonb** [comment]
@skalahonza Thanks, added [a hint](https://github.com/axuno/SmartFormat/wiki/Migration) for migrating from older versions

**Comment by axunonb** [comment]
@skalahonza @karelkral  We just published release [v3.1.0](https://github.com/axuno/SmartFormat/releases/tag/v3.1.0):
* `SmartSettings.IsThreadSafeMode = true` is the default
* Static `Smart` methods like `Smart.Format(...)` can be used also with `async` / multi-threading
